File: frontend/types/gamification.ts
import type { Achievement } from "@/types/achievement";

/**
 * Attached to any mutation response that itself triggers a
 * gamification event (analysis create/save/favorite/share, story
 * generate) — never present on a plain GET. Drives the
 * DiscoveryToast queue (app/services/gamification.py's
 * GamificationEvent is the source of truth this mirrors).
 */
export interface GamificationEvent {
  xp_awarded: number;
  total_xp: number;
  level: number;
  leveled_up: boolean;
  is_new_breed: boolean;
  is_new_rarity: boolean;
  newly_unlocked: Achievement[];
}

File: backend/app/schemas/gamification.py
from pydantic import BaseModel, Field

from app.schemas.achievement import AchievementOut


class GamificationEvent(BaseModel):
    """Attached to a mutation response (analysis create/save, favorite,
    share, story-generate) so the frontend can react to *this specific*
    action — show a toast for a level-up or newly-unlocked achievement,
    only when one genuinely just happened, without a second round-trip.

    `xp_awarded` is 0 when the underlying event had already happened
    before (e.g. re-favoriting a cat you've already favorited once) —
    idempotent by design, see CollectionEventModel.
    """

    xp_awarded: int
    total_xp: int
    level: int
    leveled_up: bool
    is_new_breed: bool = False
    is_new_rarity: bool = False
    newly_unlocked: list[AchievementOut] = Field(default_factory=list)
